nixos-render-docs: init redirects system (#353513)
NixOS/nixpkgs: pkgs/tools/nix/nixos-render-docs/README.md

# nixos-render-docs

A [CommonMark](https://commonmark.org/) and [`man-pages`](https://www.man7.org/linux/man-pages/man7/man-pages.7.html) renderer for the NixOS and Nixpkgs manuals.

## Summary

> NixOS Render Docs is a renderer originally made for documentation written in the [DocBook](https://docbook.org/whatis) format, which was then dropped in support for CommonMark, a flavour of markdown, following [this RFC](https://github.com/NixOS/rfcs/pull/72). This shift aligns the project with more accessible documentation formats and community standards (along with the recent shift to [Diátaxis](https://diataxis.fr/)), enabling easier contributions and enhancing readability for new users.

Author: But it _was_ lossy, as written in the RFC.

> NixOS Render Docs is a renderer originally made for documentation written in the [DocBook](https://docbook.org/whatis) format, which was then dropped in support for CommonMark, a flavour of markdown, following [this RFC](https://github.com/NixOS/rfcs/pull/72). This shift aligns the project with more accessible documentation formats and community standards (along with the recent shift to [Diátaxis](https://diataxis.fr/)), enabling easier contributions and enhancing readability for new users.

Reviewer: This is only half-true. The conversion was definitely loss for the Nix manual, and also broke tons of URLs and workflows. `nixos-render-docs` was tested against the DocBook output and did not loose output fidelity.

> NixOS Render Docs is a renderer originally made for documentation written in the [DocBook](https://docbook.org/whatis) format, which was then dropped in support for CommonMark, a flavour of markdown, following [this RFC](https://github.com/NixOS/rfcs/pull/72). This shift aligns the project with more accessible documentation formats and community standards (along with the recent shift to [Diátaxis](https://diataxis.fr/)), enabling easier contributions and enhancing readability for new users.

Author: Ah, okay. I have limited knowledge on this so I simply flagged it in case it was an error.

> Having our own renderer gives us a few key benefits:
> - **Control and Flexibility**: We own the code, we can manipulate and extend it however we like. An example of this is the integration of option docs and our solution to redirecting away from broken links, as described in detail below.
> - **Scalability and Extensibility**: As the NixOS documentation grows, having a renderer under our control makes it easier to scale and introduce new features without being limited by the constraints of external tools. Future enhancements, such as tested code blocks or localized versions of documentation, can be seamlessly integrated.

Reviewer: ```suggestion
Maintaining our own documentation rendering framework may appear extreme but has practical advantages:
- We never have to work around existing tools made under different assumptions
- We don't have to deal with unexpected breakage
- We can grow the framework with our evolving requirements without relying on external support or approval or the need to maintain a small diff to upstream
- The amount of code involved is minimal because it's single-purpose
```

Several alternatives were considered, a detailed analysis on the trade-offs with each alternative can be found here: https://ethercalc.net/dc4vcnnl8zv0

## Redirects system

Moving contents around can cause links to break.

Since we have our own markdown parser, we can hook into the rendering process to extract all of the metadata around each content identifier.

> The [mechanism for checking correctness of redirects](./src/nixos_render_docs/redirects.py) takes the collection of identifiers and a static mapping of redirects, validates them against a set of rules, and creates a client-side redirect mapping for each output path, as well as a `_redirects` file for server-side redirects in [Netlify syntax](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file).

Reviewer: ```suggestion
The [mechanism for checking correctness of redirects](./src/nixos_render_docs/redirects.py) takes the collection of identifiers and a mapping of the identified content to its historical locations in the output.
It validates them against a set of rules, and creates a client-side redirect mapping for each output file, as well as a `_redirects` file for server-side redirects in [Netlify syntax](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file).
```

This allows us to catch:
- Identifiers that were removed or renamed
- Content that was moved from one location to another
- Various consistency errors in the redirects mapping

### Design considerations

> The creation, movement, and removal of every identifier is captured in the Git history of the Nixpkgs repo.

Reviewer: ```suggestion
The creation, movement, and removal of every identifier is captured in the Git history.
```

> However, leveraging this information during the build process of `nixos-render-docs` is quite tricky.
> There are two key inputs to the redirect system:
> 
> 1. The current set of identifiers
> 2. A list of their historical locations
> 
> One is accessible during the build process, and the other by fetching the Git history.
> Doing both of these during the build is not feasible, and to have an external command handling the Git history would require some form of communication between it and the build process.
> This quickly gets complicated.

Reviewer: ```suggestion
However, analysing hundreds of thousands of commits during the build process is impractical.
```

Let's not go into vagueness about alternatives considered. Either spell them out or leave them out.

The chosen design is a trade-off between speed, repository size, and contributor friction:
- Stricter checks always require more attention from contributors
    - Checks should be reasonably fast and ideally happen locally, e.g. as part of the build, as anything else will substantially lengthen the feedback cycle.
- Computing redirects against previous revisions of the repository would be more space-efficient, but impractically slow.

>     - It would also require keeping an impure or ever-moving reference to those other revisions.

Reviewer: ```suggestion
    - It would also require keeping an impure or otherwise continuously updated reference to those other revisions.
```

>     - The static mapping acts like a semi-automatically updated cache that we drag along with version history
>     - Other setups, such as a dedicated service to cache a history of moved content, are more complicated and would still be impure
> - Checking in large amounts of data that is touched often, bears a risk of more merge conflicts or related build failures

Reviewer: ```suggestion
    - The static mapping acts like a semi-automatically updated cache that we drag along with version history.
    - Other setups, such as a dedicated service to cache a history of moved content, are more complicated and would still be impure.
- Checking in large amounts of data that is touched often, bears a risk of more merge conflicts or related build failures.
```

> A solution picked here is to have a static mapping of the historical locations checked into the Git tree, such that it can be read during the build process.

Reviewer: ```suggestion
The solution picked here is to have a static mapping of the historical locations checked into the Git tree, such that it can be read during the build process.
```
